# Data

I denne øvelsen vil vi se hvordan du kan manipulere datasett med `pandas`. I data science bruker vi ofte store datasett og det er derfor avgjørende å manipulere datasett på en god måte. 

I de følgende oppgavene må du alltid erstatte `___` med kode som gjør hva oppgaven spør om. 

## Pandas

Her har vi data om passasjerer på skipet Titanic og informasjon om hvorvidt de overlevde. Noen passasjergrupper hadde bedre sjanser for å overleve. I denne oppgaven skal du bruke passasjerdata for å se om du kan finne noen mønstre. Først importerer vi pandas pakken. 

In [94]:
import pandas as pd

Les inn datasettet titanic.csv i et pandas DataFrame `df`. Tips: Bruk [`pandas.read_csv`](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html). 

In [95]:
# lese inn data
df = pd.read_csv('data/titanic.csv')

Se på de første 10 radene av datasettet. Tips: Bruk [`pandas.DataFrame.head`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html). 

In [96]:
# første 10 radene
første_rad = df.head(10)
første_rad

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Beskriv alle kolonnene. For numeriske kolonner, vis i hvert fall gjennomsnitt, minimum, maximum og median. For kategoriske kolonner, vis antall unike verdier og den mest vanlige verdien. Tips: Bruk [`pandas.DataFrame.describe`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html). 

In [97]:
# data beskrivelse
beskrivelse = df.describe(include='all')
beskrivelse

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891     2   
top             NaN         NaN         NaN  Braund, Mr. Owen Harris  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383838    2.308642                      NaN   NaN   
std      257.353842    0.486592    0.836071                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

               Age       SibSp       Parch  Ticket        Fare    Cabin  \
count   714.000000  891.000000  891.000000     891  891.000000      204   
unique         NaN         NaN         NaN     681         NaN      147   
top            NaN         NaN         NaN  347082         NaN  B96 B98   
freq           NaN         NaN         NaN       7         NaN        4   
mean     29.699118    0.523008    0.381594     NaN   32.204208      NaN   
std      14.526497    1.102743    0.806057     NaN   49.693429      NaN   
min       0.420000    0.000000    0.000000     NaN    0.000000      NaN   
25%      20.125000    0.000000    0.000000     NaN    7.910400      NaN   
50%      28.000000    0.000000    0.000000     NaN   14.454200      NaN   
75%      38.000000    1.000000    0.000000     NaN   31.000000      NaN   
max      80.000000    8.000000    6.000000     NaN  512.329200      NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

Lag en ny `numpy` array med overlevelses-status (kolonnen `Survived`).  Tips: Bruk [`pandas.DataFrame.loc`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html) og [`pandas.DataFrame.to_numpy`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_numpy.html). 

In [98]:
# lage overlevelses-variabel
overlever = df.loc[:, 'Survived'].to_numpy()
overlever

array([0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1,

Regn ut total andel av pasasjerer som overlevde. Tips: Bruk [`numpy.mean`](https://numpy.org/doc/stable/reference/generated/numpy.mean.html) eller [`pandas.DataFrame.mean`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html). 


In [99]:
# andel som overlevde
andel_overlever = overlever.mean()
andel_overlever 

0.3838383838383838

Regn ut proporsjonen av kvinner og menn som overlevde. Du kan bruke [`pandas.DataFrame.groupby`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html).

In [100]:
# overlevelse blant kvinner og menn
kjonn_overlevelse = df.groupby('Sex')['Survived'].mean()
kjonn_overlevelse 

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

Regn ut proporsjonen som overlevde i hver klasse (`Pclass`). 

In [101]:
# overlevelse per klasse
klasse_overlevelse = df.groupby('Pclass')['Survived'].mean()
klasse_overlevelse

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

Lag en ny variabel i `df` med aldersgrupper [0, 18), [18, 60) og [60, 120). Tips: Du kan bruke [`pandas.cut`](https://pandas.pydata.org/docs/reference/api/pandas.cut.html).

In [102]:
# aldersgrupper
df.loc[:, 'AgeGroup'] = pd.cut(df['Age'], bins=[0,18,60,120], right=False, labels=['[0,18)', '[18,60)', '[60,120]'])

Hvor mange passasjerer er i hver aldersgruppe? Tips: Du kan bruke [`pandas.DataFrame.value_counts`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html). 

In [103]:
# passasjerer per aldersgruppe
passasjerer_per_aldersgruppe = df['AgeGroup'].value_counts()
passasjerer_per_aldersgruppe

AgeGroup
[18,60)     575
[0,18)      113
[60,120]     26
Name: count, dtype: int64

Regn ut proporsjonen som overlevde i hver aldersgruppe. 

In [104]:
# overlevelse per aldersgruppe
overlevelse_per_aldersgruppe = df.groupby('AgeGroup', observed=False)['Survived'].mean()
overlevelse_per_aldersgruppe

AgeGroup
[0,18)      0.539823
[18,60)     0.386087
[60,120]    0.269231
Name: Survived, dtype: float64

Legg til dummy-variabler for hver alderskategori. Tips: Du kan bruke [`pandas.get_dummies`](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html). 

In [105]:
# alderskategori dummy variabler

df = pd.get_dummies(df, columns=['AgeGroup'], dtype=int)
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  AgeGroup_[0,18)  \
0        0         A/5 21171   7.2500   NaN        S                0   
1        0          PC 17599  71.2833   C85        C                0   
2        0  STON/O2. 3101282   7.9250   NaN        S                0   
3        0            113803  53.1000  C123        S                0   
4        0            373450   8.0500   NaN        S                0   
..     ...               ...      ...   ...      ...              ...   
886      0            211536  13.0000   NaN        S                0   
887      0            112053  30.0000   B42        S                0   
888      2        W./C. 6607  23.4500   NaN        S                0   
889      0            111369  30.0000  C148        C                0   
890      0            370376   7.7500   NaN        Q                0   

     AgeGroup_[18,60)  AgeGroup_[60,120]  
0                   1                  0  
1                   1                  0  
2                   1                  0  
3                   1                  0  
4                   1                  0  
..                ...                ...  
886                 1                  0  
887                 1                  0  
888                 0                  0  
889                 1                  0  
890                 1                  0  

[891 rows x 15 columns]

Regn ut proporsjonen som overlevde i og utenfor aldersgruppen [18, 60) for hver kjønn. 

In [108]:
# andel som overlevde i og utenfor aldersgruppen [18, 60) for hver kjønn. 
overlevelse_kjonn_alder = df.groupby(by=["Sex", "AgeGroup_[18,60)"])["Survived"].mean()
overlevelse_kjonn_alder

Sex     AgeGroup_[18,60)
female  0                   0.696429
        1                   0.767327
male    0                   0.205882
        1                   0.179625
Name: Survived, dtype: float64

Omformater data, slik at det er tre kolonner ("PassengerId", "Variabel" og "Verdi") og inneholder data om alder og overlevelse. Det skal se slikt ut: 

|      | PassengerId | Variabel | Verdi |
|-----:|------------:|---------:|------:|
|   0  | 1           | Age      | 22.0  |
|   1  | 2           | Age      | 38.0  |
|   2  | 3           | Age      | 26.0  |
|   3  | 4           | Age      | 35.0  |
|   4  | 5           | Age      | 35.0  |
|  ... | ...         | ...      | ...   |
| 1777 | 887         | Survived | 0.0   |
| 1778 | 888         | Survived | 1.0   |
| 1779 | 889         | Survived | 0.0   |
| 1780 | 890         | Survived | 1.0   |
| 1781 | 891         | Survived | 0.0   |

Tips: Du kan bruke [`pandas.melt`](https://pandas.pydata.org/docs/reference/api/pandas.melt.html). 

In [114]:
# omformater til 
df_reformated = df.melt(id_vars='PassengerId', value_vars=['Age', 'Survived'], var_name='Variabel', value_name='Verdi')
df_reformated

PassengerId  Variabel  Verdi
0               1       Age   22.0
1               2       Age   38.0
2               3       Age   26.0
3               4       Age   35.0
4               5       Age   35.0
...           ...       ...    ...
1777          887  Survived    0.0
1778          888  Survived    1.0
1779          889  Survived    0.0
1780          890  Survived    1.0
1781          891  Survived    0.0

[1782 rows x 3 columns]